File: src/gestorAplicacion/operacional/Factura.java
package gestorAplicacion.operacional;
import java.util.ArrayList;
import java.time.LocalDateTime;

import java.util.HashMap;

import java.io.Serializable;
import java.time.*;


public class Factura implements Serializable {  
	
	//-----------SERIALIZADOR-----------
	private static final long serialVersionUID = 1L;
	
	private static ArrayList<Factura> facturas;
	static {
		facturas = new ArrayList<Factura>();
	}
	
	public static ArrayList<Factura> getFacturas() {
		return facturas;
	}

	public static void setFacturas(ArrayList<Factura> facturas) {
		Factura.facturas = facturas;
	}
	
	
	//-----------ATRIBUTOS-----------
	private int idFactura;
	private Cita cita;
	private double precioTotal;
	private LocalDateTime fecha;
	private String metodoPago;
	public static int NumFacturas=0;
	private HashMap<Producto, Integer> productosVendidos = new HashMap<Producto, Integer>(); // Dict con producto y cantidad
	
	
	//-----------CONSTRUCTORES-----------
	//-----------1. Para facturar una cita -----------
	public Factura(Cita cita, LocalDateTime fecha, String metodoPago) {
		Factura.NumFacturas++;
		this.idFactura = Factura.NumFacturas;
		this.cita = cita;
		this.fecha = fecha;
		this.metodoPago = metodoPago;
		facturas.add(this);
	}
	
	//-----------2. Para facturar productos -----------
	public Factura( HashMap<Producto, Integer> productosVendidos, LocalDateTime fechaCompra, String metodoPago) {
		Factura.NumFacturas++;
		this.idFactura = Factura.NumFacturas;
		this.fecha = fechaCompra;
		this.metodoPago = metodoPago;
		this.productosVendidos = productosVendidos;
		//this.cita.getCliente().addFactura(this);
		facturas.add(this);
	}
	
	//-----------GETTERS Y SETTERS-----------
	public int getIdFactura() {
		return this.idFactura;
	}
	public void setIdFactura(int idFactura) {
		this.idFactura = idFactura;
	}
	
	public Cita getCita() {
		return this.cita;
	}
	public void setCita(Cita cita) {
		this.cita = cita;
	}
	
	public double getPrecioTotal() {
		return this.precioTotal;
	}
	public void setPrecioTotal(double precioTotal) {
		this.precioTotal = precioTotal;
	}
	
	public LocalDateTime getFecha() {
		return this.fecha;
	}
	public void setFecha(LocalDateTime fecha) {
		this.fecha = fecha;
	}
	
	public String getMetodoPago() {
		return this.metodoPago;
	}
	public void setMetodoPago(String metodoPago) {
		this.metodoPago = metodoPago;
	}
	public HashMap<Producto, Integer> getProductosVendidos(){
		return this.productosVendidos;
	}
	public void setProductosVendidos(HashMap<Producto, Integer> productosVendidos) {
		this.productosVendidos = productosVendidos;
	}
	
	
	
	//-----------OTROS METODOS -----------
	public float precioTotalProductos() {
		float totalPagar = 0;
		for(Producto p : this.productosVendidos.keySet()) {
			
			totalPagar = totalPagar +  p.getPrecioVenta() * this.productosVendidos.get(p);
		}
		
		this.precioTotal = totalPagar;
		return totalPagar;
	}
	
	public void agregarProductosAVender(Producto producto, int cantidad) {
		this.productosVendidos.put(producto, cantidad);
	}
	
	public float precioTotalServicios() {
		float totalPagar = 0;
		ArrayList<Servicio> servicios = this.cita.getServicios();
		
		for(Servicio s : servicios) {
			totalPagar += s.getPrecio();
		}
		
		return totalPagar;
	}
	
	public float totalFactura() { //deberia ejecutarse este metodo al crear un objeto de clase Factura?
		float ans = 0;
		ans = this.precioTotalProductos()+ this.precioTotalServicios();
		this.precioTotal = ans;
		return ans;
	}
	
	
	
	
}

File: Python/src/gestorAplicacion/operacional/Factura.py
class Factura:


    #-----------ATRIBUTOS-----------
    NumFacturas =0
    _factura = []   #serialiador

    #-----------CONSTRUCTORES-----------
    #-----------1. Para facturar una cita -----------

    def __init__(self, cita, fecha, metodoPago):
        self._initialize_instance_fields()

        Factura.NumFacturas += 1
        self._idFactura = Factura.NumFacturas
        self._cita = cita
        self._fecha = fecha
        self._metodoPago = metodoPago
        self._cita.setFactura(self)
        self._cita.setEstado("Exitosa")
        Factura._factura.append(self)    #serialiador

    #-----------2. Para facturar productos -----------

    def __init__(self, productosVendidos, fechaCompra, metodoPago):
        self._initialize_instance_fields()

        Factura.NumFacturas += 1
        self._idFactura = Factura.NumFacturas
        self._fecha = fechaCompra
        self._metodoPago = metodoPago
        self._productosVendidos = productosVendidos
        #this.cita.getCliente().addFactura(this)
        Factura._factura.append(self)      #serialiador

    #-----------GETTERS Y SETTERS-----------


     @classmethod    #serialiador
    def getFacturas(cls):
        return cls._factura

    @classmethod     #serialiador
    def setFacturas(cls, facturas):
        cls._factura = facturas


    def getIdFactura(self):
        return self._idFactura
    def setIdFactura(self, idFactura):
        self._idFactura = idFactura

    def getCita(self):
        return self._cita
    def setCita(self, cita):
        self._cita = cita

    def getPrecioTotal(self):
        return self._precioTotal
    def setPrecioTotal(self, precioTotal):
        self._precioTotal = precioTotal

    def getFecha(self):
        return self._fecha
    def setFecha(self, fecha):
        self._fecha = fecha

    def getMetodoPago(self):
        return self._metodoPago
    def setMetodoPago(self, metodoPago):
        self._metodoPago = metodoPago
    def getProductosVendidos(self):
        return self._productosVendidos
    def setProductosVendidos(self, productosVendidos):
        self._productosVendidos = productosVendidos



    #-----------OTROS METODOS -----------
    def precioTotalProductos(self):
        totalPagar = 0
        for p in self._productosVendidos.keys():

            totalPagar = totalPagar + p.getPrecioVenta() * self._productosVendidos[p]

        self._precioTotal = totalPagar
        return totalPagar

    def agregarProductosAVender(self, producto, cantidad):
        self._productosVendidos[producto] = cantidad

    def precioTotalServicios(self):
        totalPagar = 0
        servicios = self._cita.getServicios()

        for s in servicios:
            totalPagar += s.getPrecio()

        return totalPagar

    def totalFactura(self):
        ans = 0
        ans = self.precioTotalProductos()+ self.precioTotalServicios()
        self._precioTotal = ans
        return ans

    def __strt__(self):
        return "Factura, id=" + str(self._idFactura) + ", " + self._cita + ", precio total=" + str(self._precioTotal) + ", fecha=" + self._fecha + ", metodo pago=" + self._metodoPago + ", productos vendidos=" + self._productosVendidos




